core/src/main/java/org/jdbi/v3/DBI.java

```diff
@@ -370,41 +382,100 @@ public void useTransaction(final TransactionIsolationLevel isolation, final Tran
+     * A convenience method which opens an extension of the given type, and yields it to a callback. A handle is opened
+     * if needed by the extention, and closed before returning to the caller.
+     *
+     * @param extensionType the type of extension
+     * @param callback      a callback which will receive the extension
+     * @param <E>           the extension type
+     * @param <X>           the exception type optionally thrown by the callback
+     * @throws NoSuchExtensionException if no {@link ExtensionFactory} is registered which supports the given extension type.
+     * @throws X                        if thrown by the callback.
      */
-    public <SqlObjectType> SqlObjectType onDemand(Class<SqlObjectType> sqlObjectType)
-    {
-        return SqlObjectBuilderBridge.onDemand(this, sqlObjectType);
+    public <E, X extends Exception> void useExtension(Class<E> extensionType, ExtensionConsumer<E, X> callback)
+            throws NoSuchExtensionException, X {
+        withExtension(extensionType, extension -> {
+            callback.useExtension(extension);
+            return null;
+        });
+    }
+
+    private class LazyHandle implements Supplier<Handle>, AutoCloseable {
+        private volatile Handle handle;
+        private volatile boolean closed = false;
+
+        public Handle get() {
+            if (handle == null) {
+                initHandle();
+            }
+            return handle;
+        }
+
+        private synchronized void initHandle() {
+            if (handle == null) {
+                if (closed) {
+                    throw new IllegalStateException("Handle is closed");
+                }
+                handle = open();
+            }
+        }
+
+        public synchronized void close() {
+            closed = true;
+            if (handle != null) {
+                handle.close();
+            }
+        }
     }
 
     /**
-     * Used to close a sql object which lacks a close() method.
-     * @param sqlObject the sql object to close
+     * Returns an extension which opens and closes handles (as needed) for individual method calls. Only public
+     * interface types may be used as on-demand extensions.
+     *
+     * @param extensionType the type of extension. Must be a public interface type.
+     * @param <E> the extension type
      */
-    public void close(Object sqlObject)
-    {
-        SqlObjectBuilderBridge.close(sqlObject);
+    public <E> E onDemand(Class<E> extensionType) throws NoSuchExtensionException {
+        if (!extensionType.isInterface()) {
+            throw new IllegalArgumentException("On-demand extensions are only supported for interfaces.");
+        }
+        if (!Modifier.isPublic(extensionType.getModifiers())) {
+            throw new IllegalArgumentException("On-demand extensions types must be public.");
+        }
+        if (!extensionRegistry.hasExtensionFor(extensionType)) {
```

Comment:
We do have `ExtensionRegistry.findExtensionFor(Class, Supplier<Handle>)` that returns an `Optional`, but that would require allocating a handle now instead of on-demand.

In this case I'm just verifying that the extension will be available when future calls to the proxy require it.